### DISCUSSION TITLE
How to handle cyclic dependencies?

### DISCUSSION THREAD

**Discussion by immerhart** [async-graphql/async-graphql#1397] [Q&A] [answered]
Thanks for the effort put in this lib!

I am new to rust and graphql, so maybe my assumptions etc. are wrong.
But as far as I understood you would define smth like this in graphql:

```
User {
  id: String,
  description: Description
}

Description {
  id: String,
  user: User
}
```

And someone could query it like:
```
User {
  id,
  Description {
    User {
      id
    }
  }
}
```
This might be a bad example but the point is that the objects are linked between each other and the API consumer can access the fields on nested levels as well (which would not work if you e.g. use an user_id instead of the User object).

In rust it is not allowed to build this cycle in a struct (I would like to use structs (with e.g. SimpleObject) for the graphql generation).
Is there some mechanic (I didn't find), which allows to get that cycle inside of graphql?

**Comment by Miaxos** [comment, marked as answer]
Hello!

In fact we do not build a really recursive struct, it's more about functions. Here is an example.

With something like this:

```rust
struct Foo;
struct Bar;

#[Object]
impl Foo {
    async fn bar(&self) -> Bar {
        Bar
    }
}

#[Object]
impl Bar {
    async fn foo(&self) -> Foo {
        Foo
    }

    async fn something(&self) -> bool {
        true
    }
}

#[Object]
impl Query {
    async fn foo(&self) -> Foo {
        Foo
    }
}

struct Query;
```

It would allow you to do query like this:

```graphql
query test4 {
  foo {
    bar {
      foo {
        bar {
          foo {
            bar {
              something
            }
          }
        }
      }
    }
  }
}
```

**Reply by immerhart** [reply]
Thank you very much, I guess if it does not work with structs, that would be a solution.
